src/box/memtx_engine.cc

```diff
@@ -1354,10 +1354,38 @@ memtx_join_f(va_list ap)
 	return rc;
 }
 
+#if defined(ENABLE_CHECKPOINT_JOIN)
+#include "memtx_checkpoint_join.c"
+#else /* !defined(ENABLE_CHECKPOINT_JOIN) */
+
+#define memtx_engine_checkpoint_join generic_engine_join
+
+int
+memtx_engine_recover_synchro(struct memtx_engine *memtx,
+			     const struct vclock *vclock,
+			     struct raft_request *raft_req,
+			     struct synchro_request *synchro_req)
+{
+	(void)memtx;
+	(void)vclock;
+	(void)raft_req;
+	(void)synchro_req;
+	diag_set(ClientError, ER_UNSUPPORTED, "Tarantool CE",
+		 "recovering synchro states from snapshot");
+	return -1;
+}
+
+#endif /* !defined(ENABLE_CHECKPOINT_JOIN) */
+
 static int
-memtx_engine_join(struct engine *engine, void *arg, struct xstream *stream)
+memtx_engine_join(struct engine *engine, void *arg,
+		  struct engine_checkpoint_cursor *cur, struct xstream *stream)
 {
-	(void)engine;
+	if (cur != NULL && cur->is_checkpoint_join)
+		/* Join from files. */
+		return memtx_engine_checkpoint_join(engine, arg, cur, stream);
```

Comment:
I'd expect `ER_UNSUPPORTED` error here in CE, not the generic join implementation, which always succeeds.

**UPDATE:** After looking at the third patch in the series, I see that you raise this error in the relay code. Splitting this change in two patches actually makes it hard to review.

Reply:
> Splitting this change in two patches actually makes it hard to review.

Hope, it's easier to review now)